presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -585,17 +596,136 @@ protected Node visitExpression(Expression node, Void context)
+            // TODO: Add many more functions
+
+            private final Map<QualifiedName, Function<Identifier, Boolean>> validateFunctions = ImmutableMap.of(
+                    AVG, this::validateAvg, APPROX_DISTINCT, this::validateApproxDistinct);
+
+            private final Map<QualifiedName, Function<Identifier, Expression>> rewriteFunctions = ImmutableMap.of(
+                    AVG, this::rewriteAvg, APPROX_DISTINCT, this::rewriteApproxDistinct);
+
+            /**
+             * Rewrites the provided argument if it is non-standard (e.g. it cannot be handled by simply re-applying function to derived columns)
+             */
+            Optional<Expression> rewriteFunction(FunctionCall functionCall)
+            {
+                if (!canRewrite(functionCall)) {
+                    return Optional.empty();
+                }
+
+                return Optional.of(rewrite(functionCall));
+            }
+
+            Expression rewrite(FunctionCall functionCall)
+            {
+                QualifiedName functionName = functionCall.getName();
+                Identifier baseTableColumn = (Identifier) functionCall.getArguments().get(0);
+
+                return rewriteFunctions.get(functionName).apply(baseTableColumn);
+            }
+
+            Expression rewriteAvg(Identifier baseTableColumn)
+            {
+                Optional<Identifier> sumDerived = getDerivedFromFunctionCall(SUM, baseTableColumn);
+                Optional<Identifier> countDerived = getDerivedFromFunctionCall(COUNT, baseTableColumn);
+                Optional<Identifier> avgDerived = getDerivedFromFunctionCall(AVG, baseTableColumn);
+
+                if (sumDerived.isPresent() && countDerived.isPresent()) { //sum(sum_res)/sum(count_res)
+                    FunctionCall sum = new FunctionCall(SUM, ImmutableList.of(sumDerived.get()));
+                    FunctionCall count = new FunctionCall(SUM, ImmutableList.of(countDerived.get()));
+                    return new ArithmeticBinaryExpression(DIVIDE, sum, count);
+                }
+                else if (sumDerived.isPresent() && avgDerived.isPresent()) { //sum(sum_res) / sum(sum_res/avg_res)
+                    FunctionCall sum = new FunctionCall(SUM, ImmutableList.of(sumDerived.get()));
+                    ArithmeticBinaryExpression reconstructedCount = new ArithmeticBinaryExpression(DIVIDE, sumDerived.get(), avgDerived.get());
+                    FunctionCall denominator = new FunctionCall(SUM, ImmutableList.of(reconstructedCount));
+
+                    return new ArithmeticBinaryExpression(DIVIDE, sum, denominator);
+                }
+
+                throw new IllegalStateException("Expected sum and average, or sum and count to be present for avg rewrite");
+            }
+
+            Expression rewriteApproxDistinct(Identifier baseTableColumn)
+            {
+                // APPROX_SET must be stored as varbinary in Hive- look for a CAST(approx_set(col) AS varbinary)
```

Comment:
Don't mention Hive in presto-main, just say that we assume approx_set is stored as varbinary